### PR TITLE
fix: DAG rendering from code package

### PR DESCRIPTION
**Pull request by saikonen**
fixes issue with rendering legacy DAG's from a code package which have a parallel foreach step.

Also cleans up the custom FlowGraph implementation to rely more on inheriting base classes from the Metaflow client

### CODE REVIEW

**Review by obgibson — COMMENTED** [review]
Confirmed that the DAG renders with the code from this branch, and fails with the code from master.

Screen Shot 2022-01-26 at 12 46 25 PM
Screen Shot 2022-01-26 at 12 46 32 PM


Is it possible to add a unit test for this?

**Comment by saikonen** [comment]
> Confirmed that the DAG renders with the code from this branch, and fails with the code from master.
> Is it possible to add a unit test for this?

Good call, added a few to cover expected behaviour and error cases  :)

File: services/ui_backend_service/data/cache/custom_flowgraph.py [hunk 1]
```diff
@@ -121,7 +16,7 @@ def visit_FunctionDef(self, node):
             self.nodes[node.name] = DAGNode(node, decos, doc if doc else '')
 
 
-class FlowGraph(object):
+class FlowGraph(_FlowGraph):
```
[the remark this answers is not in the document]
Comment (saikonen) [reply services/ui_backend_service/data/cache/custom_flowgraph.py:19]: if it makes more sense to keep a fully separate implementation of the FlowGraph for the backend then I can revert the changes for sure. Main reason for using the client-side implementation as the base class was to minimise duplication and to get some further assurance that the graph_info produced by the custom flowgraph would stay in-line schema wise with newer versions of the metaflow client (not having to maintain two completely separate implementations of the flowgraph).

[the remark this answers is not in the document]
Comment (saikonen) [reply services/ui_backend_service/data/cache/custom_flowgraph.py:19]: good point about the output_steps, will change that up. Also applies to the client-side code ;)

[the remark this answers is not in the document]
Comment (saikonen) [reply services/ui_backend_service/data/cache/custom_flowgraph.py:19]: After considering this for a while, I'm starting to lean a bit more towards keeping the separate FlowGraph implementation on the backend for now. Introducing more configuration to the existing client-side FlowGraph seems to lead to a lot of messy conditionals related to parsing, as there are a lot more limitations when traversing the AST in string-only format.

I'd still leave the deindent_docstring and DAGNode imported from the client side, as there is no difference between the implementations. So FlowGraph and StepVisitor would be maintained separate from the client. The added unit tests should catch any possible breaking changes to the imported classes in the future.

**Review by obgibson — APPROVED** [review]

**Pull request merged**